dbms/src/Storages/StorageMergeTree.cpp

```diff
@@ -233,8 +233,9 @@ void StorageMergeTree::alter(
 	if (primary_key_is_modified)
 	{
 		data.primary_expr_ast = new_primary_key_ast;
-		data.initPrimaryKey();
 	}
+	/// Reinitialize primary key because primary key column types might have changed.
+	data.initPrimaryKey();
```

Comment:
We could calculate if the PK column types have really changed and reinitialize only in that case.
But just doing it in any case is simpler and shouldn't do any harm.